File: internal/tracking/tracking.go
package tracking

import "time"

type TrackType struct {
	ID         int32     `json:"id"`
	Name       string    `json:"name"`
	Emoji      string    `json:"emoji"`
	Color      string    `json:"color"`
	MetricType string    `json:"metricType"`
	CreatedAt  time.Time `json:"createdAt"`
	UpdatedAt  time.Time `json:"updatedAt"`
}

type TrackTypeRequest struct {
	Name       string
	Emoji      string
	Color      string
	MetricType string
}

type TrackEvent struct {
	ID        int32
	Type      string
	Data      any
	CreatedAt time.Time
	UpdatedAt time.Time
}

type TrackEventRequest struct {
	Type string
	Data any
}

func fromTrackTypeRequest(request TrackTypeRequest) TrackType {
	now := time.Now()
	return TrackType{
		Name:       request.Name,
		Emoji:      request.Emoji,
		Color:      request.Color,
		MetricType: request.MetricType,
		CreatedAt:  now,
		UpdatedAt:  now,
	}
}

func fromTrackEventRequest(request TrackEventRequest) TrackEvent {
	now := time.Now()
	return TrackEvent{
		Type:      request.Type,
		Data:      request.Data,
		CreatedAt: now,
		UpdatedAt: now,
	}
}

File: ui/src/models/tracking.ts
export type TrackType = {
    id: number;
    name: string;
    emoji: string;
    color: string;
    metricType: 'this' | 'that'
    createdAt: string;
    updatedAt: string;
}

export type TrackTypeRequest = Omit<TrackType, 'id' | 'createdAt' | 'updatedAt'>

export type TrackEvent = {
    id?: number;
    type: string;
    data: unknown;
    createdAt?: string;
    updatedAt?: string;
};

export type TrackEventRequest = Omit<TrackEvent, 'id' | 'createdAt' | 'updatedAt'>

export type DefaultTrackEvent = {
    item: {
        name: string;
        emoji: string;
    };
    time: Date;
    metricType: 'this' | 'that';
    metric: number;
    createdAt?: Date;
    updatedAt?: Date;
}
